Generic Integers V2: It's Time (#3686)
rust-lang/rfcs: text/0000-generic-integers.md

- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Adds the builtin types `u<N>` and `i<N>`, allowing integers with an arbitrary size in bits.

# Motivation
[motivation]: #motivation

## Generalising code for integers

Right now, there's a *lot* of boilerplate for implementing methods for integer primitives. The standard library itself is a great example; almost the entirety of `core::num` uses some gnarly macros to define all sorts of traits for all the
integer types. One example is `Shl` and `Shr`, which are defined for not just every integer type, but every *combination* of integer types. We could easily do this with const generics instead:

```rust
impl<const N: usize, const M: usize> Shl<u<M>> for u<N> {
    type Output = u<N>;
    #[inline]
    fn shl(self, rhs: u<M>) -> u<N> {
        // implementation
    }
}
```

This will decrease compilation time across the entire Rust ecosystem, maybe not by a noticeable amount, but by some amount, due to the presence of macro-based trait implementations across the board.

## Documentation decluttering

> Having generic impls would drastically reduce the noise in the "implementations" section of rustdoc. For example, the number of implementations for `Add` for integer types really drowns out the fact that it's also implemented for strings and `std::time` types, which is useful to know too.

Reviewer: ```suggestion
Having generic impls would drastically reduce the noise in the "implementations" section of rustdoc. For example, the number of implementations for `Add` for integer types really drowns out the fact that it's also [implemented for strings](https://doc.rust-lang.org/stable/std/ops/trait.Add.html#impl-Add%3C%26str%3E-for-String) and `std::time` types, which is useful to know too.
```

## Enum optimisations

> Using a smaller number of bits for a value also has the benefit of niche enum optimisations. For example, `u<7>` represents a single ASCII character, and `Option<u<7>>` can be stored in a single byte. Additionally, `Result<u<7>, E>` also takes one byte if `E` is zero-sized.

Reviewer 2: This is not an argument, because this is already handled by pattern types ( https://github.com/rust-lang/rust/pull/120131 )

Author: These two proposals are not mutually exclusive. Both can coexist.

This works because we know that values in the range `0b0_0000000..=0b0_1111111` are valid `u<7>`, but the values in the range `0b1_0000000..=0b1_1111111` are not and can represent the zero-sized variants `None` and `Err(E)`.

## Bit masks

> Integers are very useful as a simple list of bits, and specifically for generic integers, this allows numbers of bits that aren't an existing integer type. There will probably always be a need for dedicated data structures like [`BitVec`], but at least for simple cases, being able to do this with your standard integer types is nice too.

Reviewer 2: This is bad argument, too, because this is usually handled by `bitflags` crate. You don't explain here how your approach is better than `bitflags`. So, please, start with _truly_ convincing arguments, then move to others

Reviewer 3: First, please be less dismissive of the author. Second, I don't understand why you're referencing `bitflags` here. That crate is appropriate when you have a fixed-at-design-time number of bit masks that you can sensibly give Rust identifiers to. In contrast, this section talks about the N-element bit vectors where N is may be a const parameter or it otherwise isn't sensible to give a distinct Rust identifier to each individual bit (e.g., because they correspond to `usize` indices in some other array or lanes in a SIMD vector). You can tell that this is what the RFC means because it *itself* references an existing crate that does what's being discussed here, `bitvec`, which solves a rather different problem from `bitflags`.

Author: Yeah, all of these different data structures have different meanings, and the main thing I was going for here is specifically on variable-size bit sequences. This proposal does not include any "bigint" types with arbitrary amounts of bits, and so, if you want such a structure, you're going to have to rely on the crate ecosystem for that.

`bitflags` and similar crates will still exist after this, as usual. Although it's worth noting that bitflags could be easily improved by these extra integer types, since they could easily allow niches for wrappers if the number of bits needed is less than a power of two.

```rust
pub trait MyDefault: Sized {
    fn my_default() -> Self;
}
```

Right now, with what we know, we have to copy-paste our impl for every integer type. For example, an impl for `u8` would be:

```rust
impl MyDefault for u8 {
    fn my_default() -> u8 { 1 }
}
```

Except, we'd have to replicate this for every single integer type. If we're clever, we could use a macro:

```rust
macro_rules! impl_my_default {
    ($($i: ident),*) => {$(
        impl MyDefault for $i {
            fn my_default() -> $i { 1 }
        }
    )*}
}
impl_my_default!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
```

But really, that's just making the compiler do the copy-pasting for us. Instead, we're going to use the special types `u<N>` and `i<N>` to generalise the code for us. The end result looks like:

```rust
impl<const N: usize> MyDefault for u<N> {
    fn my_default() -> u<N> { 1 }
}
impl<const N: usize> MyDefault for i<N> {
    fn my_default() -> i<N> { 1 }
}
impl MyDefault for usize {
    fn my_default() -> usize { 1 }
}
impl MyDefault for isize {
    fn my_default() -> isize { 1 }
}
```

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 4: ```suggestion
That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 32-bit system, `usize` is not the same as `u64`.
```

Or any other combination, since on a 64-bit system usize is indeed the same as u64 :smile:

Author: No, this is the correct wording.

Even though they're the same layout in memory, they're different types intentionally so that you don't need to distinguish this depending on the actual layout for the specific system.

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 5: Although now I reälize that won’t work, because `usize::BITS` is a `u32`. But then again, it might be helpful to have `::BITS_USIZE` constants anyway.

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Author: This also won't work until generic const args are stable, since associated consts aren't allowed in const generics at the moment.

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 6: > This also won't work until generic const args are stable, since associated consts aren't allowed in const generics at the moment.

no, it works fine since there are no generics in the const expression. what doesn't work is `struct S<T: Tr>(S2<{ T::ASSOC_CONST }>);`

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Author: I'm not particularly married to `usize` as a parameter, so, I'd be fine with going with `u32` given everything that's come up.

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 6: most of why I wanted `usize` as the bit width parameter type is that it's the same as arrays, and that trying to share a const generic between a `u32` bit width and an array size is basically impossible until we get casting in const generic expressions. e.g.:
```rust
pub const fn to_binary<const N: usize>(v: u<N>, buf: &mut [u8; N]) -> &str {
    let mut i = 0;
    while i < N {
        buf[i] = if (v >> (N - i - 1)) & 1 == 1 { b'1' } else { b'0' };
        i += 1;
    }
    str::from_utf8(buf).unwrap()
}
```

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 3: How often do you need an array with one element per *bit*? If the array length has any other relationship with the number of bits (e.g., adding +2 for a `0b` prefix, or the aforementioned to_le_bytes and friends), then you still need const generics expressions. And casts are possibly less problematic than most other operations, which can fail and therefore imply more post-mono errors or need some solution for propagating bounds like “N > 0”.

> That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.

Reviewer 6: replied [here](https://github.com/rust-lang/rfcs/pull/3686#issuecomment-2393410191)

The rules you'd expect apply to `u<N>` and `i<N>`, which is that `u<N>` stores values in the range `0..2.pow(N)`, and `i<N>` stores integers in the range `-2.pow(N - 1)..2.pow(N - 1)`. This means that `u<1>` only holds the values `0` and `1`, and `i<1>` only holds the values `-1` and `0`. The meaning for `u<0>` and `i<0>` is a little less clear, but they both are only allowed to contain the value `0`; the ranges end up being `0..1` and `-0.5..0.5`, which… yeah, zero is the only integer in those ranges, but it still can be confusing.

For now, if you want to ensure that your integers are the right size, you can add a `const { ... }` assertion to your implementations like so:

```rust
impl<const N: usize> MyDefault for u<N> {
    fn my_default -> u<N> {
        const { assert!(N > 1); }
        1
    }
}
```

This will cause the compiler to fail when `MyDefault` is used for `u<0>` or `u<1>`, since it will force the constant block to be evaluated. Not ideal, but it's the best we've got for now.

> ## Uncommonly sized integers

Reviewer 3: Here are two answers. Their order was chosen by an unbiased coin flip.

Obviously, the value should be negative one, because two's complement always has one more negative number than positive numbers.

Obviously, the value should be zero, because `i<N>` and `u<N>` always have this value in common.

> ## Uncommonly sized integers

Reviewer 3: Less tongue-in-cheek discussion about why/whether i0 should be permitted (in certain contexts) and what its value should be: https://discourse.llvm.org/t/rfc-support-zero-width-integers-in-the-comb-dialect-circt/78492

> ## Uncommonly sized integers

Reviewer 7: Or `i<0>` could be an uninhabitable type.

> ## Uncommonly sized integers

Reviewer 3: If it was uninhabited, every possible way of constructing an `i<N>` needs to special case N = 0, either with bounds that rule it out (not properly supported yet, and likely implies an annotation burden) or by panicking. That sounds even more inconvenient than a post-mono error.

> ## Uncommonly sized integers

Author: Not sure if you just missed this part, but I explicitly clarified the value of `i<0>` with this:

> `u<N>` are able to store integers in the range `0..2.pow(N)` and `i<N>` are able to store integers in the range `-2.pow(N-1)..2.pow(N-1)`. The cheeky specificity of "integers in the range" ensures that, for `i<0>`, the range `-0.5..0.5` only contains the integer zero; in general, `u<0>` and `i<0>` will need to be special-cased anyway, as they must be ZSTs.

This felt like the most logical approach.

> ## Uncommonly sized integers

Author: Also a side note on uninhabited types: I mention later that it's an unanswered question whether `NonZero<u<0>>` would be allowed and uninhabited, but my assumption would be yes. So, it makes sense that the weird special-casing of zero would be reserved to the types that explicitly call out zero as a special case.

> ## Uncommonly sized integers

Reviewer 8: The documentation for `NonZero` states:

> Thanks to the [null pointer optimization](https://doc.rust-lang.org/beta/std/option/index.html#representation), `NonZero<T>` and `Option<NonZero<T>>` are guaranteed to have the same size and alignment

That guarantee will be impossible to fulfill if `u/i<0>` are zero-sized.

One other side effect of having `u<N>` and `i<N>` is that we can represent a lot more types than before. For example, `u<7>` is just a seven-bit integer, which we might use to represent an ASCII character. That said, using fewer bits doesn't necessarily mean you'll use up fewer bits of memory-- for example, `u<7>` by itself still requires a `u<8>` in memory.

Overall, you should expect integers where `N` is not a power of two to take up more size/alignment than their bits might imply. There may be ways of packing the bits together in a way that optimizes the amount of space used, but once you read those values into a `u<N>` or `i<N>`, this is no longer the case.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Basic semantics

> The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.

Author: I don't mind adding this, but I'm confused how you could come to the conclusion that `i32 == i<31>` here?

> The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.

Reviewer 6: maybe if you thought `i<N>` excluded the sign bit?

> The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.

Author: Why would it, though, if `i32` is including the sign bit?

> The compiler will gain the built-in integer types `u<N>` and `i<N>`, where `const N: usize`. These be identical to existing `uN` and `iN` types wherever possible, e.g. `u<8> == u8`.

Author: I mean, I guess that makes sense, but it feels like a lot of extra steps and assumptions compared to how the type is defined. `i<0>` has zero bits, therefore, it also has no sign bit. So, the only value it's capable of representing is one where the sign doesn't matter, which is zero.

Plus, again, `i32` also has 32 bits, and we're including the sign there.

I will update the RFC when I have the time to sit down and respond to all the feedback, but I would like to understand things a bit more so I can help make sure folks aren't confused.

`usize`/`isize` continue to be separate from `u<N>`/`i<N>`, where `N` is the target pointer width, as this also applied to `uN` and `iN`. `bool` will also be made different from `u<1>` for similar reasons, even though casting `false` to `0` and `true` to `1` will remain possible.

`u<N>` are able to store integers in the range `0..2.pow(N)` and `i<N>` are able to store integers in the range `-2.pow(N-1)..2.pow(N-1)`. The cheeky specificity of "integers in the range" ensures that, for `i<0>`, the range `-0.5..0.5` only contains the integer zero; in general, `u<0>` and `i<0>` will need to be special-cased anyway, as they must be ZSTs.

> It's always valid to `as`-cast `u<N>` or `i<N>` to `u<M>` or `i<M>`, and the usual sign extension or truncation will occur depending on the bits involved. A few additional casts which are possible:

Reviewer 6: > pondering: maybe only allow `as` for the lossless & value-preserving ones? Then the truncating ones can be via a named method that makes that clearer.

you can already do `0x100u16 as u8`, so why not generalize that to all sizes for consistency? restricting `as` to lossless & value-preserving casts is a breaking change.

> It's always valid to `as`-cast `u<N>` or `i<N>` to `u<M>` or `i<M>`, and the usual sign extension or truncation will occur depending on the bits involved. A few additional casts which are possible:

Author: Yeah, the main reason is backwards compatibility here: since we can already do truncating conversions, I don't want to special-case it that you can only do that for explicitly `N = 8, 16, 32, 64, 128` and instead just allowed it everywhere. If we have a change in a future edition where only lossless conversions can be as-casted, then, we can change this behaviour, but I'd rather go with consistency for now.

> It's always valid to `as`-cast `u<N>` or `i<N>` to `u<M>` or `i<M>`, and the usual sign extension or truncation will occur depending on the bits involved. A few additional casts which are possible:

Reviewer 9: we already have Into for infallible lossless casts.

* from `bool` to `u<N>` or `i<N>`
* from `char` to `u<N>` or `i<N>`
* from `u<1>` to `bool`

> * from `u<N>` to `char`, where `N < 16`

Author: Basically, it's an extension of `u8` being castable to `char`: since that's also mostly an arbitrary cutoff point (it might be better to only say that `N <= 7` can be cast to char, since that's actually the bounds of ASCII) I decided to just make it go to the maximum possible cutoff point.

Note that casting directly to `u<0>` or `i<0>` is still allowed, to avoid forcing users to special-case their code. See later notes on possible lints.

> Integer literals can be automatically coerced to `u<N>` and `i<N>`, although generic `iN` and `uN` suffixes are left out for future RFCs. When coercing literals to an explicit `u<N>` or `i<N>`, the `overflowing_literals` lint should trigger as usual, although this should not apply for generic code. See later notes on possible lint changes.

Reviewer 6: > that default isn't listed in the documentation for `core`/`std`.

it is documented here: https://doc.rust-lang.org/1.93.1/reference/expressions/literal-expr.html#r-expr.literal.int.inference-default

> Integer literals can be automatically coerced to `u<N>` and `i<N>`, although generic `iN` and `uN` suffixes are left out for future RFCs. When coercing literals to an explicit `u<N>` or `i<N>`, the `overflowing_literals` lint should trigger as usual, although this should not apply for generic code. See later notes on possible lint changes.

Reviewer 6: > Will it just fail catastrophically like it does now, or default to 32?

I'd assume literals will default to `i32` for compatibility, if the literal is too big then that will trigger the deny-by-default overflowing literals lint. if the lint is allowed, then it will just wrap the value to fit in the literal's deduced type, just like it does now.

In general, operations on `u<N>` and `i<N>` should work the same as they do for existing integer types, although the compiler may need to special-case `N = 0` and `N = 1` if they're not supported by the backend.

> When stored, `u<N>` should always zero-extend to the size of the type and `i<N>` should always sign-extend. This means that any padding bits for `u<N>` can be expected to be zero, but padding bits for `i<N>` may be either all-zero or all-one depending on the sign.

Reviewer 10: Please clarify this to say what exactly happens when I transmute e.g. `255u8` to `u<7>` (and similar to `i<N>`). I assume it is UB, i.e., the validity invariant of these types says that the remaining bits are zero-extended / sign-extended, but the RFC should make that explicit.

Author: Yeah, I'm not quite sure what a better name is; it's the same as `rustc_layout_scalar_valid_range`, which is UB if the bits are invalid.

Author: I guess that since this is the reference description, calling them niche bits would be more appropriate? Would that feel reasonable?

Reviewer 10: No. Niche bits are an implementation detail of the enum layout algorithm, and mostly not documented.

Just describe what the valid representations of of values of this type are, i.e., what should go into [this section](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html#invalid-values) about these types.

## Limits on `N`

There are two primary limits that restrict how large `N` can be:

1. All allocations in Rust are limited to `isize::MAX` bytes.
2. Most integer methods and constants use `u32` when counting bits.

The first restriction doesn't matter since `isize::MAX` bytes is `isize::MAX * 8` bits, which is larger than `usize::MAX = isize::MAX * 2 + 1` bits.

However, the second restriction is somewhat significant: for systems where `usize::MAX > u32::MAX`, we are still effectively restricted to `N <= u32::MAX` unless we wish to change these APIs. We can treat this as effectively a post-monomorphisation error similar to the error you might get when adding very large arrays inside your type; it's unlikely that someone might encounter them, but they do exist and have to be accounted for.

It's worth noting that `u32::MAX` bits is equivalent to 0.5 GiB, and thus no integer in Rust will be able to be larger than this amount. This is seen as acceptable because at that size, people can just use their own big-integer types. For now, adding a dedicated big-integer type to the standard library is left as a potential future change.

> The compiler should be allowed to restrict `N` even further, maybe even as low as `u16::MAX`, due to other restrictions that may apply. For example, the LLVM backend currently only allows integers with widths up to `u<23>::MAX` (not a typo; 23, not 32). On 16-bit targets, using `usize` further restricts these integers to `u16::MAX` bits.

Reviewer 11: ```suggestion
The compiler should be allowed to restrict `N` even further, maybe even as low as `u16::MAX`, due to other restrictions that may apply. For example, the LLVM backend currently only allows integers with widths up to `u<23>::MAX` (not a typo; twenty three). On 16-bit targets, using `usize` further restricts these integers to `u16::MAX` bits.
```
Nit: I think spelling out _twenty three_ is more useful in this context. Having both numbers right next to each other is a bit painful to read with dyslexia (at least for me).

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Reviewer 10: You mean "using N for an array length", I assume?

Author: Yes.

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Author: Oh, it's absolutely still awkward. I guess that the justification is pretty weak here, although another one is the fact that `usize` means you don't have the extra limit at `isize::MAX` bytes for 16-bit targets.

And as always, you can easily cast the const generic in expressions, but it's harder to do so when you want to reuse it as another generic. But I get the argument that you probably won't want to reuse it anyway, and always will need some sort of computation.

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Reviewer 3: I don't think there is any realistic use case for anything close to `u<i16::MAX>`, even on 64-bit targets, so whatever integer type makes typical usage work without casts would be best. While arithmetic operations can be extrapolated to any bit width, once you're past several thousand bits a primitive integer type stops working well. Just use a "proper" bitint library, or write your bit vector ops in terms of `[u64; N]`. It's true that having a post-mono error right above N = 128 is a hassle, but there has to be *some* limit and it should be well below the point where LLVM grinds to a halt because it legalized every bitwise operation into many thousands of word-sized operations and such large basic blocks trigger pathological compile-time performance all over the place.

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Author: Yeah, I agree with you. It's actually why I mentioned that `u16::MAX` (or indeed, `i16::MAX`) is probably the maximum we should guarantee that's portable, since even at that point you're much better off with a bigint type. 32k is a ridiculously large amount of bits for one number.

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Reviewer 12: Script:
```zig
const v: uX = 0;
std.debug.print("{}", .{v});
```
Compile time:
```text
1024 6.129250764846802
2048 6.318152189254761
3072 6.924046754837036
4096 7.588716268539429
5120 10.228423833847046
6144 14.128804206848145
7168 20.459102153778076
8192 29.913185834884644
```
For reference, in a debug build of Zig without cache in my `Windows11`, the compile time increases drastically when bit width is larger than `4096`.
So I think the maximum could be even much smller for better experiement under current LLVM.

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Reviewer 9: imo 512 should be more than enough for a MVP

> While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.

Reviewer 6: Zig has 65535 as the maximum bit width, clang 18 on x86_64 has 8388608.

[`generic_const_exprs`]: https://github.com/rust-lang/rust/issues/76560

## Standard library

> The existing macro-based implementation for `uN` and `iN` should be changed to implement for only `u<N>`, `i<N>`, `usize`, and `isize` instead; this has already been implemented in a mostly-generic way and should work as expected.

Author: That's a good point; I'll add a note on using these methods for a potential clippy lint. As I think I mentioned in the RFC, clippy lints will probably be the best starting point for this since it's a good incubation ground while we're not quite sure what lints will be the most useful.

Unfortunately, there are a couple things that will have to remain implemented only for the existing powers of two due to the lack of constant bounds and complex const generics, namely:

* `From` and `TryFrom` implementations (requires const-generic bounds)
* `from_*e_bytes` and `to_*e_bytes` methods (requires [`generic_const_exprs`])

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 6: > One thing to emphasize here: getting `u128` to work was a huge endeavour, and bigger ones will be even harder for things like division -- even for 128-bit it calls out to a specific symbol for that.

LLVM has a pass specifically for expanding large divisions into a loop that doesn't use a libcall, so that shouldn't really be an issue though libcalls can still be added if you want something faster: https://github.com/llvm/llvm-project/commit/3e39b271016837425038e55e57c39e5e9b289975

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 6: as part of clang gaining support for `_BitInt(N)` where `N > 128`, basically all the work to make it work has already been done in LLVM. Div/Rem was the last missing piece and that was added in 2022.

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 3: Clang still limit `_BitInt(N)` to N <= 128 on quite a few targets: https://gcc.godbolt.org/z/8P3sMjavs

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 6: i think that's merely because they haven't got around to defining the ABI, but it all works afaict: https://llvm.godbolt.org/z/88K3ox7bh

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Author: Also worth mentioning that having `N > 128` be a post-monomorphisation error was seen as one of the biggest downsides to the previous RFC, and that this would cause more of a headache than just trying to make it work in general.

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 3: That's a good start. However, as long as Clang isn't shipping it, the people working on Clang aren't discovering and fixing any bugs specific to those platforms. The div/rem lowering happens in LLVM IR so it's hopefully pretty target-independent, but most other operations are still legalized later in the backends. That includes the operations the div/rem lowering relies on, but also any other LLVM intrinsics that the standard library uses or may want to use in the future.

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Author: I'm kind of relying a lot on the fact that even though not everyone is using `_BitInt(N)` right now, by the time we actually would be stabilising this RFC, LLVM would be a lot more robust in that regard. Kind of a role reversal from what happened with 128-bit integers: back then, we were really pushing LLVM to have better support, and C benefited from that, but now, C pushing LLVM to have better support will benefit Rust instead.

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Reviewer 3: As you say, this can be revisited later, but note that there's no guarantee that Clang will ever support `_BitInt(129)` or larger on any particular target. The C standard only requires `BITINT_MAXWIDTH >= ULLONG_WIDTH`. If some target keeps it at 128 for long enough, it could become entrenched enough that nobody wants to risk increasing it (e.g., imagine people putting stuff like `char bits[BITINT_MAXWIDTH / 8];` in some headers).

> Currently, the LLVM backend already supports generic integers (you can refer to `iN` and `uN` as much as you want), although other backends may need additional code to work with generic integers.

Author: I actually had no idea that was how the standard worked, but I shouldn't really be surprised, considering how it's C. :/

## Overflow semantics

One important factor to consider for non-power-of-two integers is that overflow will require more work than usual to account for. In particular, because we can't rely on values being truncated auto-magically when stored back in memory, we'll have to explicitly mask or shift them to ensure that the correct values are stored for the padding bits.

Because of this, the `unchecked_*` methods may actually be more important and more-often used for these integers, at least when they are not powers of two.

The compiler, or at least backends like LLVM, should be able to optimise series of operations to perform these conversions less often, but it should be noted that they must always occur, even in release mode.

## Enum variants

> For now, enum variants will still be restricted to their current set of integer types, since even [`repr(u128)`] isn't stable yet. If an RFC like [#3659] gets passed, allowing arbitrary types for enum variant tags, then `u<N>` should be included in that, although that can be added as a future extension.

Author: I actually both knew about this *and* noticed it when I reread the proposal myself due to the other comments, but I hadn't actually updated it since this is relatively low priority. I'll mark this as resolved once I do update it, though.

Note that the rest of the statement still remains: we'd need some sort of decision on arbitrary-type enum reprs, since it's not entirely clear what `repr(u<7>)` would mean. We don't allow `repr(char)` either, for example, even though such a thing *could* be properly defined if we wanted to do so.

> For now, enum variants will still be restricted to their current set of integer types, since even [`repr(u128)`] isn't stable yet. If an RFC like [#3659] gets passed, allowing arbitrary types for enum variant tags, then `u<N>` should be included in that, although that can be added as a future extension.

Reviewer 13: [~~`repr(&'static str)` when?~~](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/declarations#Enumerations-with-Cases-of-a-Raw-Value-Type)

* coercing a literal integer to a generic integer (anything besides zero might overflow without a restriction on `N`, and once restrictions on `N` become possible, we can take that into account)

Preferably, a lot of the lints surrounding generic integers should be added to clippy before being accepted into the compiler, since it's likely many of them will cause more headaches than they're worth. We can continue to investigate possible lints to avoid these errors.

# Drawbacks
[drawbacks]: #drawbacks

This is a *big* change to the language, not to be taken lightly.

One of the biggest drawbacks is that this *only* allows encoding the storage size of an integer, and doesn't let users refine integers further to only allow a range of values. For example, it would be excellent for an API taking a "percent" value to enforce that the value is between 0 and 100, but generic integers alone cannot do this.

As an example, someone might end up using `u<7>` for a percent since it allows fewer extraneous values (`101..=127`) than `u<8>` (`101..=255`), although this actually just overcomplicates the code for little benefit, and may even make the performance worse.

> Overall, things have changed dramatically since [the last time this RFC was submitted][#2581]. Back then, const generics weren't even implemented in the compiler yet, but now, they're used throughout the Rust ecosystem. Additionally, it's clear that LLVM definitely supports generic integers to a reasonable extent, and languages like [Zig] and even [C][`_BitInt`] have implemented them. A lot of people think it's time to start considering them for real.

Reviewer 10: I wouldn't say Zig has *generic* integers, it seems like they have arbitrarily-sized integers. Or is it possible to write code that is generic over the integer size?

Reviewer 14: well actually you can

```zig
const std = @import("std");

fn U(comptime bits: u16) type {
    return @Type(std.builtin.Type {
        .Int = std.builtin.Type.Int {
            .signedness = std.builtin.Signedness.unsigned,
            .bits = bits,
        },
    });
}

pub fn main() !void {
    const a: U(2) = 1;
    const b: U(2) = 3;
    // const c: U(2) = 5; // error: type 'u2' cannot represent integer value '5'
    const d = std.math.maxInt(U(147));
    std.debug.print("a={}, b={}, d={}", .{ a, b, d });
    // a=1, b=3, d=178405961588244985132285746181186892047843327
}
```

Author: I guess that example is satisfactory enough, @RalfJung? Not really sure if it's worth the effort to clarify explicitly.

Reviewer 10: Ah, neat.

C and LLVM only have concrete-width integers though, I think?

Author: I mean, C doesn't have generic *anything*, so, I guess you're right. Not 100% sure the distinction is worth it.

Reviewer 15: Clang adds `_BitInt` to C++ as an extension and the number of bits can be generic: `template <size_t N> void example(_BitInt(N) a);` will deduce `N` but it only works on the actual `_BitInt` types, not just any signed integer type.

Additionally, the names `uint` and `int` remain available for future, potentially unrelated integer types.

## Bound-based generalisation

Generalising integers over their number of bits is certainly a very logical way to generalise integers *for computers*, but generalising based upon bounds is a very natural way for humans to do it, *and* more general. For example, instead of `u<N>` and `i<N>` types, we could get away with just one type,
`int<A..=B>`. This would be more powerful than the original: for example, a percentage could be represented exactly as `int<0..=100>`. Whether an integer is signed simply depends on whether its lower bound is negative.

The primary reason for leaving this out is… well, it's a lot different from the existing integer types in the language. Additionally, such a proposal can coexist with this one, since there's nothing stopping us from making an `int<A..=B>` type in the future such that `u<N>` and `i<N>` are equivalent to the correct range of integers.

> ## Integer traits

Reviewer 3: I don't think it would be a good alternative for the standard library to provide such traits. However, they do exist in the crates.io ecosystem and are reasonably popular (mainly num-traits). Insofar the RFC is motivated by letting people generalize *their* code over the existing primitive types, it would benefit from engaging with those traits. There's definitely things a language feature can do better, e.g., `impl<N> Foo for u<N>` is less likely to cause coherence problems than `impl<T: Int> Foo for T`. On the other hand, the existing traits do solve some problems, and not all drawbacks listed in this section apply to them. They can iterate on "which methods are in which trait" and "which set of traits even exists" much more easily because semver-breaking changes are possible. It's also possible to have competing traits with overlapping sets of methods, which is mitigated by std providing inherent impls (though there can still be ambiguities: https://github.com/rust-lang/rust/issues/68826).

Previously, Rust had traits which generalized the integer types and their methods, but these were ultimately removed in favour of inherent methods. Going with a generic `u<N>` over an `Int` trait would avoid the problem of determining which methods are suited for these traits; instead, the `u<N>` type would have all of them.

Additionally, having these traits does not allow non-power-of-two `u<N>` outright, and so, this feature is a strict superset of that.

Having separate `u<N>` and `i<N>` types also solves the problem of generalising signed and unsigned integers: everything would require one impl for signed integers, and one for unsigned. This would ensure that these types have exactly the same behaviour for all `N`, only differing in the upper bound on the number of bits.

## Offering as a library

This was the main proposal last time this RFC rolled around, and as we've seen, it hasn't really worked.

> Crates like [`u`], [`bounded-integer`], and [`intx`] exist, but they come with their own host of problems:

Reviewer 3: As far as I can tell, `bounded-integer` and `intx` only provide subsets of the native types up to {i,u}128, not arbitrarily large fixed-size integers. The `u` crate seems to be about something else entirely, did you mean to link something different there?

So where are the libraries that even try to do what this RFC proposes: arbitrary number of bits, driven by const generics? I've searched and found [ruint](https://docs.rs/ruint/latest/ruint/), which appears relevant.

Author: That definitely seems like a good option to add to the list. I had trouble finding them, so, I appreciate it.

Reviewer 16: I'd appreciate a mention of `https://crates.io/crates/arbitrary-int`, which is (I think) the closest in design to this rfc

> * None of these libraries can easily unify with the existing `uN` and `iN` types.

Reviewer 3: A const-generic library type can't provide this and also can't support literals. But what problems exactly does that cause in practice? Which aspects can be handled well with existing language features and which ones really need language support?

Author: The RFC already mentions how being able to provide a small number of generic impls that cover all integer types has an extremely large benefit over being forced to use macros to implement for all of them individually. You cannot do this without language support.

Reviewer 3: So this bullet point is "only" about impls like `impl<const BITS: usize> Foo for some_library::Int<BITS> { ... }` not implementing anything for the primitive integer types? Could `From` impls and some form of delegation (#3530) also help with this?

Author: Not really, and this is mentioned in the RFC also. That's 5 impls for unsigned, 5 impls for signed that could just be 2 impls, whether you have delegation or not. Even for simple traits, like `Default`, you're incentivised to use a macro just because it becomes so cumbersome.

Reviewer 16: `arbitrary-int` provides a unification somewhat using its `Number` trait. It's somewhat rudimentary but I am working on improving it.

Reviewer 16: Reading this again, the Number trait fulfills a somewhat different role though. It allows writing generic code against any Number (be it an arbitrary-int or a native int), but it does not expose the bits itself - which can be a plus or a minus, depending on what you're building.

> * Generally, they require a lot of unsafe code to work.

Reviewer 3: What kind of unsafe code, and for what purposes? And is that sufficient reason to extend the language? Usually, if it's something that can be hidden behind a safe abstraction once and for all, then it seems secondary whether that unsafety lives on crates.io, in sysroot crates, or in the functional correctness of the compiler backend.

Author: Generally, the unsafe code is stuff similar to the `bounded-integer` crate, where integers are represented using enums and transmuted from primitives. The casting to primitives is safe, but not the transmuting.

Reviewer 3: Is that really all? Because that seems trivial to encapsulate without affecting the API, and likely to be solved by any future feature that makes it easier to opt into niche optimizations (e.g., pattern types).

Author: Yeah, it's easy to encapsulate, but I think it's worth mentioning that unsafe code is involved as a negative because it means many code bases will be more apprehensive to use it.

You are right that it could easily be improved, though, with more compiler features. I just can't imagine it ever being on par with the performance of a compiler-supported version, both at runtime and compile time.

Reviewer 16: arbitrary-int works without unsafe code (with the exception of the optional function `new_unchecked` which skips the bounds check)

Reviewer 2: @clarfonthey
@hanna-kruppe
> likely to be solved by any future feature that makes it easier to opt into niche optimizations (e.g., pattern types)

And this is already solved by pattern types in nightly: https://play.rust-lang.org/?version=nightly&mode=release&edition=2024&gist=583297ef353ad5e8e79907c06c3c197f

> * These representations tend to be slower and less-optimized than compiler-generated versions.

Reviewer 3: Do we have any data on what's slower and why? Are there any lower-stakes ways to fix these performance issues by, for example, adding/stabilizing suitable helper functions (like https://github.com/rust-lang/rust/issues/85532) or adding more peephole optimizations in MIR and/or LLVM?

Author: Main source of slowdown is from using enums to take advantage of niche optimisations; having an enum with a large number of variants to represent this niche is pretty slow to compile, even though most of the resulting code ends up as no-ops after optimisations.

I definitely should mention that I meant slow to compile here, not slow to run. Any library solution can be made fast to run, but will generally suffer in compile time when these features are effectively *already* supported by the compiler backends, mostly for free.

Reviewer 3: Is there any compile time issue when *not* trying to provide niches? Out of the potential use cases the RFC lists, only a couple seem to really care about niche optimizations. In particular, I don't expect that it typically matters for integers larger than 128 bits. (But again, surveying the real ecosystem would help!) If so, the compile time problem for crates like bounded-integer could be addressed more directly by stabilizing a proper way to directly opt into niches instead of having to abuse enums. And that would help with *any* bounds, while this RFC (without future possibilities) would not.

Reviewer 3: Well, I would expect some negative compile-time impact from repeatedly monomorphizing code that's const-generics over bit width or bounds. But that's sort of inherent in having lots of code that is generic in this way, so it's no worse for third party libraries than for something built-in.

Author: That's very fair; I agree that we should have an ability to opt into niches regardless. I guess that my reasoning here is pretty lackluster because I felt that the other reasons to have this feature were strong enough that this argument wasn't worth arguing, although you're right that I should actually put a proper argument for it.

From what I've seen, of the use cases for generic integers:

1. Generalising primitives
2. Between-primitives integer types (like `u<7>` and `u<48>`)
3. Larger-than-primitives integer types

For 1, basically no library solution can work, so, that's off the table. For 2, which is mostly the subject of discussion here, you're right that it could probably be improved a lot with existing support. And for 3, most people just don't find the need to make *generalised* code for their use cases, and just explicitly implement, say, `u256` themselves with the few operations they need.

The main argument IMHO is that we can effectively knock out all three of these options easily with generic integers supported by the language, and they would be efficient and optimized by the compiler. We can definitely whittle down the issues with 2 and 3 as we add more support, but the point is that we don't need to if we add in generic integers.

Although, I really need to solidify this argument, because folks like you aren't 100% convinced, and I think that the feedback has been pretty valuable.

Reviewer 3: Yeah, I appreciate that you're trying to tackle a lot of different problems with a unifying mechanism. I focus on each problem separately because I want to tease out how much value the unifying mechanism adds for each of them, compared to smaller, more incremental additions that may be useful and/or necessary in any case. Only when that's done I feel like I can form an opinion on whether this relatively large feature seems worth it.

> * They still require you to generalise integer types with macros instead of const generics.

Reviewer 3: I'm not sure I understand the problem here. If a library provides `struct Int<const BITS: usize>(...);` then code using this library shouldn't need macros to interact with it (except, perhaps, as workaround for current gaps in const generics). The library itself would have a bunch of impls relating its types to the language primitives, which may be generated with macros. But that doesn't seem like such a drastic problem, if it's constrained to the innards of one library, or a few competing libraries.

Author: I'm not sure I understand your argument. No matter what, a library solution cannot be both generic *and* unify with the standard library types. I don't see a path forward that would allow, for example, some library `Uint<N>` type to allow `Uint<8>` being an alias for `u8` while also supporting arbitrary `Uint<N>`. Even with specialisation, I can't imagine a sound subset of specialisation allowing this to work.

Like, sure, a set of libraries can choose to only use these types instead of the primitives, circumventing the problem. But most people will want to implement their traits for primitives for interoperability.

Reviewer 3: This overlaps a bit with the bullet point about unification, but I do think it depends a lot on what one is doing. For example, the num-traits crate defines traits that it needs to implement for the primitive types. On the other hand, any code that's currently written against the traits from num-traits may be happy with a third party library that provides `Int<N>` and `Uint<N>` and implements the relevant traits for them. And for something like bit fields, you may not need much interop with the primitive types at all: in the `MipsInstruction` example, you probably want some widening and narrowing conversions, but only with respect to u32 specifically.

It's hard to form an opinion about how common these scenarios are (and whether there are other nuances) without having a corpus of "real" code to look at. Experience reports (including negative ones) with crates like num-traits and bounded-integer may be more useful than discussing it in the abstract.

A library solution really doesn't feel like the right option here. While libraries can create general integer *traits* to work over all of the existing `uN` and `iN` types, they can't easily make generic integer types.

[`u`]: https://docs.rs/u
[`bounded-integer`]: https://docs.rs/bounded-integer
[`intx`]: https://docs.rs/intx

## Going without

This is always an option, but hopefully it seems like a worse option after all that's been said so far.

# Prior art
[prior-art]: #prior-art

* [The previous RFC.][#2581]
* [Zulip RFC revival topic.][Zulip]
* [Generic integers in Zig.][Zig]
* [Generic integers in C23.][`_BitInt`]

> * Probably several others discussions I'm missing.

Reviewer 17: is the recent [swift](https://forums.swift.org/t/integer-generic-parameters/74181) proposal relevant here?

Author: Probably! Thank you for letting me know; I'll add it to the list.

Author: Actually, no, it's not; I totally should have read that before updating the RFC. It appears to just be an RFC for const generics, which Rust already has, and not generic integers.

Reviewer 17: apologies, I agree that it actually is only const generics, rather than these generic integers.... oops

Author: All good! Any blame I could assign to you for not reading I could also assign to myself, so…
